### ISSUE TITLE
[Request] llama.cpp

### ISSUE THREAD

**Issue by jlo62** [chaotic-aur/packages#3193]
### Package:

https://aur.archlinux.org/packages/llama.cpp
https://aur.archlinux.org/packages/llama.cpp-cuda
https://aur.archlinux.org/packages/llama.cpp-hip
https://aur.archlinux.org/packages/llama.cpp-opencl
https://aur.archlinux.org/packages/llama.cpp-sycl-f16
https://aur.archlinux.org/packages/llama.cpp-sycl-f32
https://aur.archlinux.org/packages/llama.cpp-vulkan

### Purpose:

llama.cpp is a port of Facebook's llama model in C/C++, and supports running many large language models (llms).

### Benefits:

These packages contain gpu accelerated builds, which can speed up runtime by a multiple

### Building:

~~The git release is useful for following upstream, as the non-git versions are often outdated and (at least amds) often lag behind/are unusable~~

### Copyright:

MIT

### Expected Interest:

Many

### Already available?

No

### Unique request?

Yes

### Banned package?

No

**Comment by Technetium1** [comment]
Can be added when it compiles successfully: [llama.cpp-git.log](https://builds.garudalinux.org/repos/chaotic-aur/logs/llama.cpp-git.log)

**Comment by xiota** [comment]
Looks like a library that no other package currently uses.

**Comment by Technetium1** [comment]
@xiota runs as an interactive program as well as a server.

The AUR package may need to be updated due to https://github.com/ggerganov/llama.cpp/pull/7809

**Comment by txtsd** [comment]
I've created a set of packages that you might be interested in including.  
These should all build correctly in a chroot since I use `pkgctl build` to test.  

[llama.cpp](https://aur.archlinux.org/packages/llama.cpp)  
[llama.cpp-vulkan](https://aur.archlinux.org/packages/llama.cpp-vulkan)  
[llama.cpp-sycl-fp16](https://aur.archlinux.org/packages/llama.cpp-sycl-f16)  
[llama.cpp-sycl-fp32](https://aur.archlinux.org/packages/llama.cpp-sycl-f32)  
[llama.cpp-opencl](https://aur.archlinux.org/packages/llama.cpp-opencl)  
[llama.cpp-cuda](https://aur.archlinux.org/packages/llama.cpp-cuda)  
[llama.cpp-hip](https://aur.archlinux.org/packages/llama.cpp-hip)  

Also, should I be posting this here or on gitlab?

**Comment by xiota** [comment]
@txtsd Here is okay.  Why did you make each package separate instead of part of the same pkgbuild, like the git package?

**Comment by jlo62** [comment]
>instead of part of the same pkgbuild, like the git package?

This is the main reason I'd like to see the builds here, building all 7 if I only need one is annoying.

@txtsd :+1:

**Comment by xiota** [comment]
@jlo62 How long do builds take (individual and combined)? How big are the packages?

**Comment by txtsd** [comment]
@xiota

- Maintenance burden is too high if it's all in one package. If one fails to build, I have to restart building without cache all over again (`pkgctl build`).
- Dependencies are too large (25GB+). I don't have enough space on my `/` to install deps for all at once.
- With `pkgctl build`, copying deps to the chroot as part of the build process takes way too long.

So I split it up into separate packages.

I don't remember how long they take to build, but the -hip and -cuda take the longest.

Packages sizes:

```
.rw-r--r-- 4.5M txtsd txtsd 26 Oct 21:05 llama.cpp/llama.cpp-b3982-1-x86_64.pkg.tar.zst
.rw-r--r--  92M txtsd txtsd 27 Oct 04:05 llama.cpp-cuda/llama.cpp-cuda-b3982-2-x86_64.pkg.tar.zst
.rw-r--r--  84M txtsd txtsd 27 Oct 04:06 llama.cpp-cuda/llama.cpp-cuda-debug-b3982-2-x86_64.pkg.tar.zst
.rw-r--r--  84M txtsd txtsd 26 Oct 21:06 llama.cpp/llama.cpp-debug-b3982-1-x86_64.pkg.tar.zst
.rw-r--r--  11M txtsd txtsd 27 Oct 01:19 llama.cpp-hip/llama.cpp-opencl-b3982-2-x86_64.pkg.tar.zst
.rw-r--r-- 4.5M txtsd txtsd 27 Oct 01:34 llama.cpp-opencl/llama.cpp-opencl-b3982-2-x86_64.pkg.tar.zst
.rw-r--r--  43M txtsd txtsd 27 Oct 01:19 llama.cpp-hip/llama.cpp-opencl-debug-b3982-2-x86_64.pkg.tar.zst
.rw-r--r--  84M txtsd txtsd 27 Oct 01:35 llama.cpp-opencl/llama.cpp-opencl-debug-b3982-2-x86_64.pkg.tar.zst
.rw-r--r-- 6.6M txtsd txtsd 26 Oct 23:39 llama.cpp-sycl-f16/llama.cpp-sycl-f16-b3982-1-x86_64.pkg.tar.zst
.rw-r--r--  45M txtsd txtsd 26 Oct 23:40 llama.cpp-sycl-f16/llama.cpp-sycl-f16-debug-b3982-1-x86_64.pkg.tar.zst
.rw-r--r-- 6.6M txtsd txtsd 26 Oct 22:18 llama.cpp-sycl-f32/llama.cpp-sycl-f32-b3982-1-x86_64.pkg.tar.zst
.rw-r--r--  45M txtsd txtsd 26 Oct 22:19 llama.cpp-sycl-f32/llama.cpp-sycl-f32-debug-b3982-1-x86_64.pkg.tar.zst
.rw-r--r-- 5.0M txtsd txtsd 27 Oct 01:38 llama.cpp-vulkan/llama.cpp-vulkan-b3982-1-x86_64.pkg.tar.zst
.rw-r--r--  88M txtsd txtsd 27 Oct 01:39 llama.cpp-vulkan/llama.cpp-vulkan-debug-b3982-1-x86_64.pkg.tar.zst
```

Build folder sizes:

```
λ du -h --summarize llama.cpp*
1.4G	llama.cpp
798M	llama.cpp-cuda
446M	llama.cpp-hip
1.4G	llama.cpp-opencl
1.3G	llama.cpp-sycl-f16
1.4G	llama.cpp-sycl-f32
1.5G	llama.cpp-vulkan
```

**Comment by xiota** [comment]
@txtsd Thank you for the info.  I'll add one at a time and let you know if I run into any problems.

**Comment by xiota** [comment]
@jlo62 I'm switching this request to the stable version because maintainer is accessible and responsive.

**Comment by xiota** [comment]
@dr460nf1r3 Problem with `.` in package name? https://gitlab.com/chaotic-aur/pkgbuilds/-/pipelines/1515413743

**Comment by txtsd** [comment]
> Package names can contain only alphanumeric characters and any of @, ., _, +, -. Names are not allowed to start with hyphens or dots. All letters should be lowercase.

From: https://wiki.archlinux.org/title/Arch_package_guidelines#Package_naming

Should be accounted for if that's what it is :eyes:

**Comment by xiota** [comment]
@txtsd It's a problem with builder, not packaging.  Someone who isn't me (or you) has to look at it.

**Comment by txtsd** [comment]
I know. I was just clarifying :smile_cat:

**Comment by dr460nf1r3** [comment]
> @dr460nf1r3 Problem with `.` in package name? https://gitlab.com/chaotic-aur/pkgbuilds/-/pipelines/1515413743

This didn't detect any changes in the folder at all, therefore I'm quite sure this has nothing to do with the dot 🤔 the detection checks each folder besides the ones starting with a dot for any changes.

**Comment by dr460nf1r3** [comment]
https://github.com/chaotic-cx/chaotic-repository-template/commit/a1eadf243ca5f29dc5c47b72c234d2cd297cbeab

**Comment by xiota** [comment]
Not sure that regex is working.  See https://gitlab.com/chaotic-aur/pkgbuilds/-/pipelines/1517191414

Does `bash` regex recognize `\w`?  This doesn't work in shell.
```bash
if [[ "a" =~ ^([\w]+) ]]; then
  echo true
fi
```

Maybe something like this would work:
```bash
local _chars='a-z0-9@_+'
if [[ "$file" =~ ^([${_chars}][${_chars}\.-]*)/ ]]; then
  : do stuff
fi
```

**Comment by dr460nf1r3** [comment]
> Not sure that regex is working.  See https://gitlab.com/chaotic-aur/pkgbuilds/-/pipelines/1517191414
> 
> Does `bash` regex recognize `\w`?  This doesn't work in shell.
> ```bash
> if [[ "a" =~ ^([\w]+) ]]; then
>   echo true
> fi
> ```
> 
> Maybe something like this would work:
> ```bash
> local _chars='a-z0-9@_+'
> if [[ "$file" =~ ^([${_chars}][${_chars}\.-]*)/ ]]; then
>   : do stuff
> fi
> ```

I tested it via regex101 and `\w` is a quite common key.. hm. I now wonder what flavor of regex bash uses 🤔
If you want, you can PR the change to the template repo. Will be working today but can merge stuff.

**Comment by xiota** [comment]
Is there a gitlab instance or open a PR at [github.com/chaotic-cx/chaotic-repository-template](https://github.com/chaotic-cx/chaotic-repository-template) ?

**Comment by dr460nf1r3** [comment]
It's the GitHub repo only. Though I already amended my commit.

**Comment by xiota** [comment]
Leave this at `llama.cpp` for now.  Added interfere (571518779fa3ad6fca467ec3cba0e1e3fe6a9686) to use `libggml-git`.  Needed to avoid conflict with `whisper.cpp`.  Using [custom package](https://github.com/chaotic-aur/packages/tree/main/libggml-git) because AUR package has some problems, like illegal instruction error.

**Comment by txtsd** [comment]
@xiota Just to be sure, you're asking me to:

1. Build `llama.cpp*` in a generic manner by using the build options from `libggml-git`
2. Remove the built libggml and use `libggml-git` as a dependency to avoid a conflict with `whisper.cpp`?

**Comment by xiota** [comment]
@txtsd 

The interfere fixes the issues with `llama.cpp` as far as chaotic-aur repo is concerned.

1. Would be best practice to make `llama.cpp` work with baseline x86_64, using the build options from the custom libggml-git package in this repo.

2. Whether to remove the libggml files after build is up to you.  I personally would not unless a user asks.  However, since the variants seem to work by changing libggml build parameters, maybe there should be only one version of `llama.cpp` with multiple versions of `libggml`.  (Same may apply to `whisper.cpp`.)

You could also try linking libggml statically so that there are no conflicting files in the final package.  This might require me to rewrite the interfere.

**Comment by txtsd** [comment]
@xiota I made the `llama.cpp*` packages use baseline x86_64 using the custom `libggm-git` package as a reference. Please see if it's sufficient, and let me know.


I'm not yet sure how to link libggml statically. I tried toggling the `GGML_STATIC` flag, but that just builds and bundles a `libggml.a`. I'm not sure if that's correct.

**Comment by xiota** [comment]
@txtsd I don't have hardware to test all the configurations, but the changes look okay to me.

I haven't tested with these specific packages, but usually, after static linking, `.a`, headers, and other related files can be deleted.

**Comment by txtsd** [comment]
I will investigate further, thanks!

**Comment by xiota** [comment]
@txtsd I took a look at `whisper.cpp`, and I think static builds would probably be best.  If static build cannot be done, could install to alternate location (`/usr/lib/llama.cpp`) and set rpath.

If you want help, you can add me as comaintainer of `llama.cpp`.  Once that one is working, it should be usable as a model for the others.

**Comment by txtsd** [comment]
@xiota Thanks for your offer of help. I made all the packages build statically.

You were right, removing the `.a` files was the way to go. I was under the impression `makepkg` handles that internally with the default `!staticlibs` option.

I tested with a model and prompt to make sure `llama.cpp` worked before I ported the changes to the other packages.

**Comment by txtsd** [comment]
We might have a big problem with statically built libggml in llama.cpp (at least for CUDA)

See: https://aur.archlinux.org/packages/llama.cpp-cuda#comment-998638

**Comment by xiota** [comment]
Switched to a custom package.  Package name uses `-cpp` instead of `.cpp`.  Currently not possible to use a separately built copy of `libggml` without significant changes to the build scripts.  So the easiest way to prevent conflicts is to use an isolated directory `/usr/lib/llama-cpp/`.  Binaries are not symlinked.  Add `/usr/lib/llama-cpp/bin` to path if desired.

Users who need specific acceleration support should build the relevant AUR package themselves.

This package may be dropped if there is not sufficient usage.

**Comment by xiota** [comment]
Added `vulkan-x64v3` variant by request.  Dropped `devel` packages because utility isn't clear.  If someone needs them, please briefly describe what they're being used for.  Metrics:

* llama-cpp-git # 113
* llama-cpp-devel-git # 23
* llama-cpp-vulkan-git # 128
* llama-cpp-vulkan-devel-git # 34
* llama-cpp-vulkan-x64v3-git # 1

**Comment by alexjp** [comment]
Did some tests using the new `x64v3` package.
The more layers in CPU I use, the bigger the difference.

Example benchmark result:
* First one is using `llama-cpp-vulkan-x64v3-git`
* Second one is using `llama-cpp-vulkan-git`
* Third one is again retesting with `llama-cpp-vulkan-git` to be sure

![Image](https://github.com/user-attachments/assets/e2dfdab7-7d41-44d4-9a31-f777e459b470)


TL;DR:

In this example test, it went from 2.85 tokens per second to 8.41!

**Comment by xiota** [comment]
Seems that vulkan builds work without it, so moved vulkan support into the main package.  Also added a separate cuda build.  Metrics:

* llama-cpp-git # 133
* llama-cpp-x64v3-git # 11
* llama-cpp-cuda-git # 6
* llama-cpp-vulkan-git # 99 (dropped)
* llama-cpp-vulkan-x64v3-git # 46 (dropped)

**Comment by alexjp** [comment]
> Seems that vulkan builds work without it, so moved vulkan support into the main package. Also added a separate cuda build. Metrics:
> 
> * llama-cpp-git # 133
> * llama-cpp-x64v3-git # 11
> * llama-cpp-cuda-git # 6
> * llama-cpp-vulkan-git # 99 (dropped)
> * llama-cpp-vulkan-x64v3-git # 46 (dropped)

Noticed the change a few days back, everything seems to be working fine! (certainly vulkan still working)

**Comment by xiota** [comment]
Revised again... because...  Everyone knows `ld.so --help` shows supported x86-64 micro-architectures, but it also says "Subdirectories of glibc-hwcaps directories ... supported, searched..."  What does that mean?

glibc can load optimized libraries if it finds a structure like:
```
/usr/lib/library.so
/usr/lib/glibc-hwcaps/x86-64-v2/library.so
/usr/lib/glibc-hwcaps/x86-64-v3/library.so
/usr/lib/glibc-hwcaps/x86-64-v4/library.so
```

So in exchange for extra disk space and compilation time, libraries optimized for different micro-architectures can be stuffed into a single package.  This was known [since around 2021](https://bbs.archlinux.org/viewtopic.php?id=263371), but still no packages do this?

Metrics:

* llama-cpp-git # 146
* llama-cpp-x64v3-git # 112
* llama-cpp-cuda-git # 137

**Comment by xiota** [comment]
Based on benchmarks, will remove v2 and v4 in a future revision.  Leaving them for now so other people can try them.

I ran v1 and v2 twice.  Results were basically the same with v2 performing slightly worse than v1.

When Vulkan is enabled, v1, v3, v4 performed about the same (pp512 around 1130 t/s).  v2 seemed to have some sort of performance regression (pp512 around 895-1000 t/s).  Results were fairly consistent across several runs.

```
$ # x86-64-v4
$ GLIBC_TUNABLES="glibc.cpu.hwcaps=" /opt/llama-cpp/bin/llama-bench -ngl 0 --cpu-strict 1 -b 256 -m llama-3.2-1b-instruct-q8_0.gguf
| llama 1B Q8_0                  |   1.22 GiB |     1.24 B | Vulkan,BLAS,RPC |       8 |          1 |     256 |           pp512 |         96.80 ± 0.62 |
| llama 1B Q8_0                  |   1.22 GiB |     1.24 B | Vulkan,BLAS,RPC |       8 |          1 |     256 |           tg128 |         21.04 ± 0.03 |

$ # x86-64-v3
$ GLIBC_TUNABLES="glibc.cpu.hwcaps=-AVX512F" /opt/llama-cpp/bin/llama-bench -ngl 0 --cpu-strict 1 -b 256 -m llama-3.2-1b-instruct-q8_0.gguf
| llama 1B Q8_0                  |   1.22 GiB |     1.24 B | Vulkan,BLAS,RPC |       8 |          1 |     256 |           pp512 |         96.36 ± 1.26 |
| llama 1B Q8_0                  |   1.22 GiB |     1.24 B | Vulkan,BLAS,RPC |       8 |          1 |     256 |           tg128 |         20.94 ± 0.08 |

$ # x86-64-v2
$ GLIBC_TUNABLES="glibc.cpu.hwcaps=-AVX2" /opt/llama-cpp/bin/llama-bench -ngl 0 --cpu-strict 1 -b 256 -m llama-3.2-1b-instruct-q8_0.gguf
| llama 1B Q8_0                  |   1.22 GiB |     1.24 B | Vulkan,BLAS,RPC |       8 |          1 |     256 |           pp512 |         73.64 ± 0.91 |
| llama 1B Q8_0                  |   1.22 GiB |     1.24 B | Vulkan,BLAS,RPC |       8 |          1 |     256 |           tg128 |         18.93 ± 0.10 |

$ # x86-64
$ GLIBC_TUNABLES="glibc.cpu.hwcaps=-SSE4_2" /opt/llama-cpp/bin/llama-bench -ngl 0 --cpu-strict 1 -b 256 -m llama-3.2-1b-instruct-q8_0.gguf
| llama 1B Q8_0                  |   1.22 GiB |     1.24 B | Vulkan,BLAS,RPC |       8 |          1 |     256 |           pp512 |         74.99 ± 1.07 |
| llama 1B Q8_0                  |   1.22 GiB |     1.24 B | Vulkan,BLAS,RPC |       8 |          1 |     256 |           tg128 |         18.35 ± 0.03 |
```

The selected microarchitecture level can be confirmed:
```
GLIBC_TUNABLES="glibc.cpu.hwcaps=-AVX2" ld.so --help | grep x86-64-v
  x86-64-v4
  x86-64-v3
  x86-64-v2 (supported, searched)
```

**Comment by xiota** [comment]
Revised yet again.  Rebuild pending.  This time, the ggml library is moved to a separate package (`libggml-git`) because upstream added option to build with system library.  `libggml-git` is a custom package with openblas, vulkan, and v1/v3 optimizations, using the glibc-hwcaps mechanism previously described.
